aips/on-chain-randomness-impl.md

```diff
@@ -0,0 +1,419 @@
+Status: Draft
+last-call-end-date (*optional): <mm/dd/yyyy the last date to leave feedbacks and reviews>
+type: Standard (Core, Networking, Framework)
+created: 02/12/2024
+updated (*optional): <mm/dd/yyyy>
+requires (*optional): <AIP number(s)>
+---
+
+# AIP-X - Implementation of instant on-chain randomness
+
+## Summary
+
+This AIP describes the implementation of the randomness API from AIP-41[^aip-41], focusing on the cryptographic constructions and their efficient integration into our system. Specifically, we describe how we use a weighted distributed key generation (DKG) protocol in a PoS setting to set up threshold keys among the validators, and have validators generate randomness every block using a weighted verifiable unpredictable function (VUF). We will also describe other required changes such as reconfiguration change.
+
+### Goals
+
+The goal is to provide Move smart contracts with access to (1) **instant**[^aip-41], (2) **unbiasable** and (3) **unpredictable** randomness w.r.t. to any minority stake of the validators.
+
+At the same time, we must do this without imposing a high penalty on the blockchain’s latency or throughput.
+
+## Motivation
+
+As already described in AIP-41[^aip-41] the need for on-chain randomness arises in many applications, such as decentralized games, raffles, randomized NFTs, randomized airdrops, and more. Plus, many blockchain applications will benefit from randomness in terms of fairness, security, and functionality.
+
+## Impact
+
+On-chain randomness interacts with and impacts our blockchain protocol. The two protocols must be carefully co-designed to maintain performance, especially in terms of latency.
+
+## Alternative solutions
+
+**VDFs.** Current techniques based on verifiable delay functions (VDFs) introduce high latencies in the worst paths (such as [Bicorn](https://eprint.iacr.org/2023/221)). This does not meet the [goals](#Goals) described above.
+
+**Interactive VSS and VUFs**. A different design would use a DKG based on interactive VSS to deal a secret amongst our validators. Unfortunately, this design requires that new-epoch validators be online when setting up their threshold keys for randomness generation, which does not align with our current system architecture.
+
+**Other insecure approaches.** Several existing blockchains have insecure on-chain randomness implementations, as described in our [blog post](https://aptoslabs.medium.com/roll-with-move-secure-instant-randomness-on-aptos-c0e219df3fb1). 
+
+## Specification
+
+On-chain randomness introduces several changes to the current Aptos blockchain system, including running a DKG for every epoch change, modifying the existing reconfiguration process, and generating a **randomness seed** for every block. This section describes the above system changes at a high level. Implementation details are discussed in [Reference Implementation](#reference-implementation).
```

Comment:
can we add in who is running or who can run the dkg? is this a service hosted by aptos labs?

Reply:
done